src/package-hoister.js

```diff
@@ -291,8 +326,24 @@ export default class PackageHoister {
       }
     }
 
+    const peerDependencies = Object.keys(info.pkg.peerDependencies || {});
+
     // remove redundant parts that wont collide
-    while (parts.length) {
+    hoistLoop: while (parts.length) {
+      // we must not hoist a package higher than its peer dependencies
+      for (const peerDependency of peerDependencies) {
+        const checkParts = parts.concat(peerDependency);
+        const checkKey = this.implodeKey(checkParts);
+        info.addHistory(`Looked at ${checkKey} for a peer dependency match`);
+
+        const existing = this.tree.get(checkKey);
+
+        if (existing) {
+          info.addHistory(`Found a peer dependency requirement at ${checkKey}`);
+          break hoistLoop;
```

Comment:
I don't know - despite being usually avoided, the label has a great use case here, and I feel like the hoister is complex enough to warrant some of its logic to be kept inline, so that we don't have to jump everywhere in the file to understand its process.